docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,529 @@
+    return new_params, MomentumState(new_momentum)
+```
+
+Remarks:
+
+- We'll call `Optimizer` instances from the functional API `optim`, and refer
+  to `flax.optim.Optimizer` instances from the [previous API] as `optimizer`
+  for the remainder of this document.
+- As opposed to the [previous API], this API does not have the option to update
+  hyper parameters. When [using Optax], updating the hyper parameters is instead
+  done by defining a schedule that takes care of updating the hyper parameters
+  during training.
+- Using this simple API directly makes the user code a bit more verbose than
+  using the [previous API], but the difference is smaller with Linen. If you
+  care about a boilerplate-free training loop see the proposed [Linen helper]
+  further down.
+
+## Using Optax
+[Using Optax]: #using-optax
+
+Optax is centered on the idea of composable gradient transformations. Since
+above [functional API] is based on Optax's interface it is straightforward to
+wrap an Optax transformation to comply with the `Optimizer` interface:
+
+```python
+import optax
+
+
+def optax_optimizer(tx: optax.GradientTransformation) -> Optimizer:
+  """Wraps an Optax transformation to update parameters directly."""
+
+  def update_fn(
+    grads: Gradients, params: Params, state: OptimizerState
+  ) -> Tuple[Params, OptimizerState]:
+    updates, state = tx.update(grads, state, params)
+    params = optax.apply_updates(params, updates)
+    return params, state
+
+  return Optimizer(tx.init, update_fn)
+
+tx = optax.chain(
+    optax.trace(decay=0.9, nesterov=False),
+    optax.scale(-0.1),
+)
+
+optim = optax_optimizer(tx)
+```
+
+The only difference is that the Optax udpate function returns a transformed
+gradient and the updated optimizer state (which makes them composable), so we
+apply this to the parameters in a final step and return the updated parameters
+and optimizer state.
+
+
+## Multi Optimizer
+[Multi Optimizer]: #multi-optimizer
+
+The [previous API] defined `flax.optim.MultiOptimizer` for processing different
+parts of the parameter tree with different optimizers:
+
+```python
+biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: path.endswith('/bias'))
+not_biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: not path.endswith('/bias'))
+
+optimizer_def = flax.optim.MultiOptimizer(
+    (biases_traversal, flax.optim.GradientDescent(learning_rate=0.1)),
+    (not_biases_traversal, flax.optim.GradientDescent(learning_rate=0.05)),
+)
+```
+
+Note how we first define a traversal that selects parameters based on their
+path (which is the concatenation of module scopes and variable name), and then
+create a `MultiOptimizer` that binds a different optimizer for each of these
+separate traversals.
+
+A similar scheme could be implemented with optimizers following the functional
+API:
+
+```python
+# flax/opt/base.py
+
+FilterFn = Callable[[str], bool]
+
+
+def multi_optimizer(
```

Comment:
I'm not sure if this is the idiomatic pattern in `optax`, but there is a approved [PR](https://github.com/deepmind/optax/pull/27) on a mask wrapper to update only a subset of the parameters in a pytree. I think this is a clean way to apply different optimizers to different parameters within an `optax.chain`.

Reply:
Now that the PR is merged, here's how a `multi_optimizer` _could_ look in pure Optax:

```python
def dict_map_with_path(fn, d, path=''):  # a crude "ModelParamTraversal"
    if isinstance(d, dict):
        return {k: dict_map_with_path(fn, v, path+'/'+k) for k, v in d.items()}
    return fn(path, d)

# bool masks with same structure as (or a prefix of) the params dict
is_bias_mask = dict_map_with_path(
    lambda path, value: path.endswith('/bias'), params)
not_bias_mask = jax.tree_map(lambda x: not x, is_bias_mask)

multi_optimizer = optax.chain(
    optax.masked(optax.sgd(learning_rate=0.1), mask=is_bias_mask),
    optax.masked(optax.sgd(learning_rate=0.05), mask=not_bias_mask))
```

Here's a [colab gist](https://colab.research.google.com/gist/n2cholas/2c36724f1d740ed9d388681c41f6e479/optax-multioptimizer.ipynb) with a full example. 

@mtthss is this the idiomatic approach, or is there a better way?

Reply:
I do like the approach supported by @n2cholas's recently contributed `masked` wrapper,
because it makes no assumption on how you may want to specify the `mask`
(e.g. it compatible with flax/stax/haiku/etc users to each generate the mask however is more idiomatic for them)
but would be good to hear from flax's multiopt users whether it works for them

@avital what do you think about this?

Reply:
Seems reasonable to me, and as far as I can tell, this is just as powerful as our https://flax.readthedocs.io/en/latest/flax.optim.html#flax.optim.ModelParamTraversal. (See the docstring [here](https://flax.readthedocs.io/en/latest/flax.optim.html#flax.optim.MultiOptimizer) for an example usage)

But one thing I don't like so much about this optax mask API is that it represents paths as "full strings" (IIUC, or maybe I am misunderstanding?). Can we instead make them lists of (nested) keys?

Reply:
I am probably missing something, but I am not sure what you mean by "it represents paths as full strings"?

Right now the mask is a pytree of booleans with the same structure as the params (or a prefix thereof)
So for instance a very simple mask that just looks as shapes would be:
```
mask = jax.tree_util.tree_map(lambda x: x.ndim != 1, params)
custom_weight_decay = optax.masked(optax.add_decayed_weights(0.001), mask)
```

Could you elaborate on what make them "lists of nested keys" would look like?

I am not particularly "married" to the current mask representation,
my main desiderata would be:
1) simple to do simple things
2) flexible enough to do complicated things
3) nn library agnostic

We can certainly choose a different one and switch to it if you think there is a more appropriate one.

Reply:
Oh I see -- this design allows `optax.masked` to bypass the question of "how to represent nested keys" by asking you to tree_map yourself, I now understand and I like it. So there remains the questions of how one would make a mask that depends on keys, e.g. "only weights, not biases".

@n2cholas shared a utility function above, `dict_map_with_path` that allows for that but that implementation encodes nested keys as "key1/key2/key3", that is what I was referring to. I prefer that we make this function exist somewhere general (maybe in optax for now, but should be in some "jax commons" place) but with a slightly different implementation:

```
def dict_map_with_path(fn, d, path=()):
    if isinstance(d, dict):
        return {k: dict_map_with_path(fn, v, path + (k, )) for k, v in d.items()}
    return fn(path, d)
```

this would allow us to suggest a pretty easy transition plan for users of `flax.optim.ModelParamTraversal` (the `filter_fn` argument to `ModelParamTraversal` could be fed directly into `dict_map_with_path`)

Reply:
(deleted my previous comment because I misremembered how `ModelParamTraversal` worked -- it does represent the argument passed to `filter_fn` as string e.g. '/Conv_0/weight', not tuples/lists). 

Updated, correct comment below:

Oh I see -- this design allows optax.masked to bypass the question of "how to represent nested keys" by asking you to tree_map yourself, I now understand and I like it. So there remains the questions of how one would make a mask that depends on keys, e.g. "only weights, not biases".

@n2cholas shared a utility function above, `dict_map_with_path` that allows for that but that implementation encodes nested keys as "key1/key2/key3", that is what I was referring to. To make the smoothest For people who currently use Flax's `ModelParamTraversal`, we'd want to use something very similar to that. The signature of the `filter_fn` argument to the `ModelParamTraversal` constructor is very similar to what you'd pass to `dict_map_with_path` (Flax also adds a leading /)

So your proposal sounds like a good plan.

Reply:
@avital I just want to clarify `dict_map_with_path` was very ad-hoc and just for the sake of example. I like the idea of this function using a tuple for the path, as this matches the format of Flax's [`flatten_dict`](https://flax.readthedocs.io/en/latest/flax.traverse_util.html#flax.traverse_util.flatten_dict)

Haiku also has some nice utilities including a `haiku.data_structures.map` function that provides similar information (docs [here](https://dm-haiku.readthedocs.io/en/latest/api.html#haiku.data_structures.map)).

Reply:
Cool, updated the FLIP accordingly.

Note though that now we need to have `params` ready already when we define the optimizer - previously this was deferred to calling `optim.init()` - I don't think it makes a difference for our examples, though.

Reply:
Note: with https://github.com/deepmind/optax/pull/60 above comment about `optim.init()` does not apply anymore.